introduce `develop` section (#253)
compose-spec/compose-spec: develop.md

# The Compose Specification - Development support
{:.no_toc}

*Note:* Development support is an OPTIONAL part of the Compose Specification.

* ToC
{:toc}

## Introduction

Compose specification is a platform-neutral way to define multi-container applications. Compose focusing on development
use-case to run application on local machine also supports some development hooks to improve developers velocity on their
local workflow (also known as "inner loop"). This document defines how Compose behaves to efficiently assist the developer.

## Definitions

Compose Specification supports an OPTIONAL `develop` subsection on services. This section defines the development
constraints and workflows to be set by Compose to assist the developer working on service codebase. Only a subset of
Compose file services may define such a `develop` subsection.

## Illustrative sample

The following sample illustrates Compose specification concepts with a concrete sample application:

```yaml
services:
  frontend:
    image: awesome/webapp
    build: ./webapp
    develop:
      watch: 
        # sync static content

>         - path: ./src
>           action: sync
>           target: /var/www

Reviewer: Build context is `./webapp`, so the path should probably be under it.

```suggestion
        - path: ./webapp/src
          action: sync
          target: /app
```

          ignore:
            - node_modules/

  backend:
    image: awesome/backend
    build: ./backend
    develop:
      watch: 
        # rebuild image and recreate service

>         - path: ./src

Reviewer: ```suggestion
        - path: ./backend
```

Compose to monitor source code for changes.

##### action

`action` define the action to take place as changes have been detected:

- `rebuild` strategy will rebuild service image based on the `build` section and recreate the service with updated image.
- `sync` strategy keep existing service container(s) running, but synchronize source files with container content according to `target` attribute. 

##### ignore

The `ignore` attribute can be used to define a list of patterns for paths to be ignored. Any updated file
that matches a pattern, or belongs to a folder that matches a pattern, won't trigger service to be re-created. 
The syntax is the same as `.dockerignore` file: 

- `*` matches 0 or more characters in a file name. 
- `?` matches a single character in file name. 
- `*/*` matches two nested folders with arbitrary names
- `**` matches an arbitrary number of nested folders

If the build context includes a `.dockerignore` file, the patterns in this file are loaded as implicit content
for the `ignores` file, and values set in the Compose model are appended.

##### path

> `path` attribute defines the path to source code (relative to `build.context`) to monitor for changes. Updates to any file

Reviewer: ```suggestion
`path` attribute defines the path to source code (relative to the project directory) to monitor for changes. Updates to any file
```

Author: Really ? When did we change this ?

inside path, which doesn't match any `ignore` rule, will trigger configured action.

##### target

> `target` attribute onbly applies when `action` is configured for `sync`. Files within `path` with changes will be synchronized

Reviewer: ```suggestion
`target` attribute only applies when `action` is configured for `sync`. Files within `path` with changes will be synchronized
```
